### PR TITLE
General improvements, resolve alpha/beta comparisons

### PR DESCRIPTION
**Pull request by ExtremeFiretop**
Pleas see Gnuton PR 571 and 600 for the reason we started investigating this: 

PR: https://github.com/gnuton/asuswrt-merlin.ng/pull/571
PR: https://github.com/gnuton/asuswrt-merlin.ng/pull/600

image

This is our original issue which brought this up to our attention: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/issues/147

### CODE REVIEW

File: release/src/router/rom/webs_scripts/merlin_webs_update.sh [hunk 1]
```diff
@@ -93,11 +93,11 @@ if [ "$webs_state_flag" -eq "1" ]; then
 	if [ "$forsq" == "1" ]; then
 		echo "---- download SQ release note $fwsite/test/$releasenote_file0_US ----" >> /tmp/webs_upgrade.log
 		/usr/sbin/wget $wget_options $fwsite/test/$releasenote_file0_US -O $releasenote_path0
-		echo "---- $fwsite/test/$releasenote_file0 ----" >> /tmp/webs_upgrade.log
+		echo "---- $fwsite/test/$releasenote_file0_US ----" >> /tmp/webs_upgrade.log
```
Comment (ExtremeFiretop) [inline release/src/router/rom/webs_scripts/merlin_webs_update.sh:96]: Note that $releasenote_file0 is not a valid variable
It should be $releasenote_file0_US

File: release/src/router/rom/webs_scripts/merlin_webs_update.sh [hunk 2]
```diff
@@ -18,7 +18,7 @@ fi
 current_base=$(nvram get firmver | sed "s/\.//g")
 current_firm=$(nvram get buildno | cut -d. -f1)
 current_buildno=$(nvram get buildno | cut -d. -f2)
-current_extendno=$(nvram get extendno | sed "s/-g.*//" | sed "s/_.*//" | sed "s/alpha\|beta/-1/")
+current_extendno=$(nvram get extendno | awk -F'[_-]' '{n=$1} /[aA]lpha|[bB]eta/{n--} END{print n}') #Extract extendno, subtract value by 1 if it contains "alpha/beta", remove all other values such as "_rog" or "-g*"
```
Comment (ExtremeFiretop) [inline release/src/router/rom/webs_scripts/merlin_webs_update.sh:21]: This has never properly worked...
# Current solution:

Define the extendnos as a space-separated string:
```
extendnos="2_alpha1-g4889cee45a 0 2_alpha3 3 0_rog 0_alpha1 0_rog_alpha1 1_alpha1 1_rog_alpha1 0_Alpha1 0_Rog_Alpha1 0_Beta2 1_beta3"
```
Loop through each extendno:
```
for extendno in $extendnos; do
  current_extendno=$(echo "$extendno" | sed "s/-g.*//" | sed "s/_.*//" | sed "s/alpha\|beta/-1/")
  echo "Original: $extendno -> Processed: $current_extendno"
done
```

# Output:
Original: 2_alpha1-g4889cee45a -> Processed: 2
Original: 0 -> Processed: 0
Original: 2_alpha3 -> Processed: 2
Original: 3 -> Processed: 3
Original: 0_rog -> Processed: 0
Original: 0_alpha1 -> Processed: 0
Original: 0_rog_alpha1 -> Processed: 0
Original: 1_alpha1 -> Processed: 1
Original: 1_rog_alpha1 -> Processed: 1
Original: 0_Alpha1 -> Processed: 0
Original: 0_Rog_Alpha1 -> Processed: 0
Original: 0_Beta2 -> Processed: 0
Original: 1_beta3 -> Processed: 1

Comment (ExtremeFiretop) [reply release/src/router/rom/webs_scripts/merlin_webs_update.sh:21]: # Proposed solution:

Define the extendnos as a space-separated string
```
extendnos="2_alpha1-g4889cee45a 0 2_alpha3 3 0_rog 0_alpha1 0_rog_alpha1 1_alpha1 1_rog_alpha1 0_Alpha1 0_Rog_Alpha1 0_Beta2 1_beta3"
```

Loop through each extendno
```
for extendno in $extendnos; do
  # Process each extendno in a single line
  processed_number=$(echo "$extendno" | awk -F'[_-]' '{n=$1} /[aA]lpha|[bB]eta/{n--} END{print n}')
  echo "Original: $extendno -> Processed: $processed_number"
done
```

# Output:
Original: 2_alpha1-g4889cee45a -> Processed: 1
Original: 0 -> Processed: 0
Original: 2_alpha3 -> Processed: 1
Original: 3 -> Processed: 3
Original: 0_rog -> Processed: 0
Original: 0_alpha1 -> Processed: -1
Original: 0_rog_alpha1 -> Processed: -1
Original: 1_alpha1 -> Processed: 0
Original: 1_rog_alpha1 -> Processed: 0
Original: 0_Alpha1 -> Processed: -1
Original: 0_Rog_Alpha1 -> Processed: -1
Original: 0_Beta2 -> Processed: -1
Original: 1_beta3 -> Processed: 0

File: release/src/router/rom/webs_scripts/merlin_webs_update.sh [hunk 3]
```diff
@@ -37,13 +37,13 @@ if [ "$?" != "0" ]; then
...
 
-	fullver=$(grep $model /tmp/wlan_update.txt | sed s/.*#FW//;)
+	fullver="$(grep "$model" /tmp/wlan_update.txt | tail -n1 | sed 's/.*#FW//')"
 	fullver=$(echo $fullver | sed s/#.*//;)
 	firmbase=$(echo $fullver | cut -d. -f1)
 	firmver=$(echo $fullver | cut -d. -f2)
 	buildno=$(echo $fullver | cut -d. -f3)
 
-	extendno=$(grep $model /tmp/wlan_update.txt | sed s/.*#EXT//;)
+	extendno="$(grep "$model" /tmp/wlan_update.txt | tail -n1 | sed 's/.*#EXT//')"
```
Comment (ExtremeFiretop) [inline release/src/router/rom/webs_scripts/merlin_webs_update.sh:46]: To properly tail variables with unstandard Input.
saves the script from "Input cmd" when using -rog, -g4889cee45a, etc

Comment (ExtremeFiretop) [reply release/src/router/rom/webs_scripts/merlin_webs_update.sh:46]: Examples found in https://github.com/gnuton/asuswrt-merlin.ng/pull/571

File: release/src/router/rom/webs_scripts/merlin_webs_update.sh (hunk 1 above)
Comment (RMerl) [reply release/src/router/rom/webs_scripts/merlin_webs_update.sh:96]: SQ selease is unused in production, it's old QA/testing code that allows us developers to temporarily point to a different manifest.

**Comment by RMerl** [comment]
I am confused by this PR, as there are comments scattered all over the place.  What is the actual issue this commit is addressing?

File: release/src/router/rom/webs_scripts/merlin_webs_update.sh (hunk 1 above)
Comment (ExtremeFiretop) [reply release/src/router/rom/webs_scripts/merlin_webs_update.sh:96]: @RMerl 

I see, In such a case, I would normally comment out the code
It's just not the standard I'm accustomed to, since the variable will still be read and considered invalid whether it's used or not.

For debugging purposes, usually we comment out code.

**Comment by ExtremeFiretop** [comment]
> I am confused by this PR, as there are comments scattered all over the place. What is the actual issue this commit is addressing?

Sorry, hopefully you can make sense of it if you read here: https://github.com/RMerl/asuswrt-merlin.ng/pull/851/files
It's really only 3 small adjustments and the reasoning for them is already explained.

> Pleas see Gnuton PR 571 and 600 for the reason we started investigating this:
> 
> PR: [gnuton/asuswrt-merlin.ng#571](https://github.com/gnuton/asuswrt-merlin.ng/pull/571) PR: [gnuton/asuswrt-merlin.ng#600](https://github.com/gnuton/asuswrt-merlin.ng/pull/600)
> 
> This is our original issue which brought this up to our attention: [ExtremeFiretop/MerlinAutoUpdate-Router#147](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/issues/147)

But if it helps I'll go copy and paste the discussions of value from those topics for you and add them here.

**Comment by ExtremeFiretop** [comment]
> I am confused by this PR, as there are comments scattered all over the place. What is the actual issue this commit is addressing?

The changes to these lines:

`extendno=$(grep $model /tmp/wlan_update.txt | sed s/.*#EXT//;)`

Is to match the changes done for Gnuton, the changes were just to be replicated upstream.
The issue in question is here:
image

The issue in specific here: https://github.com/gnuton/asuswrt-merlin.ng/pull/571 as mentioned.

The changes to these lines:

`current_extendno=$(nvram get extendno | sed "s/-g.*//" | sed "s/_.*//" | sed "s/alpha\|beta/-1/")
`

Is to address this issue reported:

https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/issues/147

TL;DR

my best guess of the issue, is that alpha version wasn't following the normal standards for versioning, and so the webs_update.sh script is failing. Because our script piggy-backs on the routers built-in update checker, if this script fails, our script will never detect a newer release available. 

That points to the built-in update script failing to parse the version installed correctly.
I already know you may say Betas or at least Alphas, don't follow a standard for versioning. But in that case that's more of a use case to have a the webs_update.sh be more universally adaptable.

That's the thing, Alphas & Betas can be wild cards WRT a "standard" format (perhaps Alphas more than Betas).
All we can do is best-effort attempts to catch them. This this that attempt to catch them.

Changes to:
`echo "---- $fwsite/test/$releasenote_file0 ----" >> /tmp/webs_upgrade.log`

Was due to the fact that we put the script in Gntuons script he had it set to -x (xtrace)
We changed it to -u (nounset) so we could see invalid variables for troubleshooting, while doing so we identified the fact that the console is reading these unset variables even if they aren't used.

In your case, that's not going to stop the script, but that is why the PR is called "General improvements"

**Comment by ExtremeFiretop** [comment]
@RMerl 

Let me know Eric if you have issues making sense of anything else. :)

**Comment by RMerl** [comment]
I'm currently working on something else so I don't have the time to go through all past discussions.  But here is some context.

This script is a hacked and patched together script based on a 15+ years old update script from Asus.  At the time I only focussed on making it work with my own manifest, the original script was quite messy.  It included support for alternate manifests (the SQ flag), also for international changelogs (hence the _US and the non _US reference), etc...  Asus ultimately totally abandonned that script, reimplementing it as a closed source function, so I just left things as they were as the script was doing what it was meant to do for me: retrieve the manifest, look for an update, wget the changelog and set the nvram flags.  Making only necessary changes as needed, such as adding support for the 3004/3006 kernel version.

SQ mode: that flag is controlled through the "apps_sq" nvram.  Asus betatesters used it to have the router automatically retrieve notifications for internal beta releases, in addition to accessing beta versions of Optware/Downloadmaster.  I left it there only for the very rare occasions where I needed to test a manifest format change, which may have happened twice in 10+ years.  A more appropriate fix for that section would be to drop support for SQ manifests, cleaning up the whole code section.

release note variable: the whole echo to the update log can be removed, as there is no real use for this (since unlike Asus, I don't provide regionalized changelogs, so I don't need to have the actual filename logged).

alpha/beta checks: the string has always been lower case, so I don't see the need to start looking for both lower and upper-cased instances.  Final releases always properly generated a new firmware notification for me when my router was running a beta version.

Handling _rog or other special flags: aside from the fact that _rog will be dropped in 3006, do you mean that the update script failed to detect new versions when the router is running an _rog version of the firmware?

**Comment by RMerl** [comment]
Actually, logigng the changelog filename might still be worth keeping, since the filename is derived from the version string.

**Comment by RMerl** [comment]
Commit 90a9cc3a53e0415056949e0eec6318bb28244c9a addresses the following:

- Get rid of legacy SQ support
- FIx logged changelog filename
- Fix detection of whether wget succeeded or not (echo call was overwriting the wget result)

File: release/src/router/rom/webs_scripts/merlin_webs_update.sh (hunk 1 above)
Comment (ExtremeFiretop) [reply release/src/router/rom/webs_scripts/merlin_webs_update.sh]: @RMerl 

Resolving as your commit: https://github.com/RMerl/asuswrt-merlin.ng/commit/90a9cc3a53e0415056949e0eec6318bb28244c9a deals with it in a cleaner manner than just patching the variable.

**Comment by ExtremeFiretop** [comment]
> Commit [90a9cc3](https://github.com/RMerl/asuswrt-merlin.ng/commit/90a9cc3a53e0415056949e0eec6318bb28244c9a) addresses the following:
> 
> * Get rid of legacy SQ support
> * Fix logged changelog filename
> * Fix detection of whether wget succeeded or not (echo call was overwriting the wget result)

@RMerl 

Confirmed on my side, thanks for that commit. I resolved the noted variable change notes since your PR cleans it up nicely compared to just changing the variable to a valid one.

> I'm currently working on something else so I don't have the time to go through all past discussions. But here is some context.
> 

Eric, don't feel like you need to respond to things in a rushed manner, take your time, and respond to me when you have a second. Nothing here is urgent. :)

> alpha/beta checks: the string has always been lower case, so I don't see the need to start looking for both lower and upper-cased instances. Final releases always properly generated a new firmware notification for me when my router was running a beta version.
> 

I think my PR was to address the compatibility of alphas better more than beta's; while yes I did include the upper and lower (Just to be safe) the change is really so that when the extendno contains "alpha" it will properly subtract the corresponding extendno number for the comparison further down.

> Handling _rog or other special flags: aside from the fact that _rog will be dropped in 3006, do you mean that the update script failed to detect new versions when the router is running an _rog version of the firmware?

Correct, there are instances were the script failed to detect an update due to the special flags, examples I could think off were such as **_rog**  or **_alpha1-g234567**, the example I linked too was from Gnuton's script, but they are literally almost identical except for the URL and the setting of -u (nounset). This helps with rog support on current versions, which I know are being dropped in 3006, (Gnuton plans to continue to support TUF and ROG UIs in 3006) but also for alphas.

Now while I understand that alphas aren't technically supported as I mentioned here:

> "resolve alpha/beta comparisons" is technically an issue in code, whether you support them or not.

I just figured I'd try to address better alpha support while I was in edit mode.
My understanding of the original line here:

`current_extendno=$(echo "$extendno" | sed "s/-g.*//" | sed "s/_.*//" | sed "s/alpha\|beta/-1/")`

Was that if it contained *alpha* or *beta* then it should subtract the extendno so it can be properly compared here:

`&& [ "$current_extendno" -lt "$lextendno" ]; then`

And be considered "lower" than production releases of the same version.

However as I identified in my review notes above, for that part of the code, that never seemed to work for me or the user which reported the issue on our MerlinAU github page:

Original: **2_alpha1**-g4889cee45a -> Processed: **2**
Original: **2_alpha3** -> Processed: **2**
Original: **0_alpha1** -> Processed: **0**
Original: **0_rog_alpha1** -> Processed: **0**
Original: **1_alpha1** -> Processed: **1**
Original: **1_rog_alpha1** -> Processed: **1**
Original: **0_Alpha1** -> Processed: **0**
Original: **0_Rog_Alpha1** -> Processed: **0**

The change was so that it would properly subtract the extendno if it contained alpha before processing/stripping the rest of the string.

Original: **2_alpha1**-g4889cee45a -> Processed: **1**
Original: **2_alpha3** -> Processed: **1**
Original: **0_alpha1** -> Processed: **-1**
Original: **0_rog_alpha1** -> Processed: **-1**
Original: **1_alpha1** -> Processed: **0**
Original: **1_rog_alpha1** -> Processed: **0**
Original: **0_Alpha1** -> Processed: **-1**
Original: **0_Rog_Alpha1** -> Processed: **-1**

Example was here: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/issues/147

**Comment by ExtremeFiretop** [comment]
Actually, I just tested, and the ROG continues to work (at least for me.) So it's fully about the alphas/matching Gnuton in this case :)

**Comment by ExtremeFiretop** [comment]
@RMerl 

I will close this and open a new one that better reflects the outstanding issues :)

**Comment by ExtremeFiretop** [comment]
New PR opened here: https://github.com/RMerl/asuswrt-merlin.ng/pull/853
